Add OPNSense device tracker

mtreinish opened the pull request:
## Description:

This commit adds a new component for using an OPNSense router as a
device tracker. It uses pyopnsense to query the api to look at the
arptable for a list of devices on the network.


**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10427

## Example entry for `configuration.yaml` (if applicable):
```yaml
opnsense:
   url: http://$router_ip/api
   api_key: api_key
   api_secret: api_secret
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [X] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

mtreinish committed:
Add OPNSense device_tracker

This commit adds a new component for using an OPNSense router as a
device tracker. It uses pyopnsense to query the api to look at the
arptable for a list of devices on the network.

mtreinish committed:
Run black formatting locally to appease azure

fabaff commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,72 @@
+    url = conf[CONF_URL]
+    api_key = conf[CONF_API_KEY]
+    api_secret = conf[CONF_API_SECRET]
+    verify_ssl = conf.get(CONF_VERIFY_SSL, False)
```
There is already a default from the validation.

mtreinish replied:
Removed in: https://github.com/home-assistant/home-assistant/pull/26834/commits/914fce5e613221f69c8fcb7edbcf20a97b0c418d

fabaff commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        for interface in tracker_interfaces:
+            if interface not in interfaces:
+                _LOGGER.error(
+                    "Specified OPNsense tracker interface %s is not " "found", interface
```
```suggestion
                    "Specified OPNsense tracker interface %s is not found", interface
```

mtreinish replied:
Oops, good catch. Thanks

mtreinish replied:
Ah, it looks like black did this. That's why I was confused that i missed it :)

fabaff commented on homeassistant/components/opnsense/manifest.json:
```diff
@@ -0,0 +1,10 @@
+    "pyopnsense==0.2.0"
+  ],
+  "dependencies": [],
+  "codeowners": []
```
Please add yourself as code owner.

mtreinish replied:
Oh, I just assumed I didn't need to because I'm not listed as a codeowner on any of the components I've contributed in the past. Figured it was a requirement. Fixed in: https://github.com/home-assistant/home-assistant/pull/26834/commits/914fce5e613221f69c8fcb7edbcf20a97b0c418d

fabaff commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,72 @@
+
+    interfaces_client = diagnostics.InterfaceClient(
+        api_key, api_secret, url, verify_ssl
+    )
```
Dito

mtreinish replied:
Same as above, the client classes all share the same base class in pyopnsense.

mtreinish replied:
Fixed by: https://github.com/home-assistant/home-assistant/pull/26834/commits/3f762edbf417ac6154d9216eadbfed1193d88bf9

fabaff commented on homeassistant/components/opnsense/device_tracker.py:
```diff
@@ -0,0 +1,67 @@
+        self.client = client
+        self.interfaces = interfaces
+
+    def _get_mac_addrs(self, devices):
```
Docstring is missing.

mtreinish replied:
Added in: https://github.com/home-assistant/home-assistant/pull/26834/commits/914fce5e613221f69c8fcb7edbcf20a97b0c418d

fabaff commented on homeassistant/components/opnsense/device_tracker.py:
```diff
@@ -0,0 +1,67 @@
+
+
+class OPNSenseDeviceScanner(DeviceScanner):
+    """This class queries a router running ASUSWRT firmware."""
```
Stale docstring.

mtreinish replied:
Oops, fixed in: https://github.com/home-assistant/home-assistant/pull/26834/commits/914fce5e613221f69c8fcb7edbcf20a97b0c418d

fabaff commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        # Verify that specified tracker interfaces are valid
+        netinsight_client = diagnostics.NetworkInsightClient(
+            api_key, api_secret, url, verify_ssl
+        )
```
I assume that there will be a traceback if the credentials are wrong or the device is not present.

mtreinish replied:
Hmm, actually it doesn't fail until it is actually used (which to be fair is the next line). The init just stores the passed in values: https://github.com/mtreinish/pyopnsense/blob/master/pyopnsense/client.py#L34-L39

MartinHjelmare replied:
We should return False from component setup if credentials are bad.

mtreinish replied:
Fixed by: https://github.com/home-assistant/home-assistant/pull/26834/commits/3f762edbf417ac6154d9216eadbfed1193d88bf9

fabaff commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        api_key, api_secret, url, verify_ssl
+    )
+    clients = {"interfaces": interfaces_client}
+    hass.data[OPNSENSE_DATA] = clients
```
`clients` is only used once. From my point of view it's clear from reading the code but it's your call.

```suggestion
    hass.data[OPNSENSE_DATA] = {"interfaces": interfaces_client}
```

fabaff commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,72 @@
+    verify_ssl = conf.get(CONF_VERIFY_SSL, False)
+    tracker_interfaces = conf.get(CONF_TRACKER_INTERFACE, None)
+
+    from pyopnsense import diagnostics
```
Can now be at the top of the file.

mtreinish replied:
Ooh, things have changed since my last contribution :) Moved in: https://github.com/home-assistant/home-assistant/pull/26834/commits/914fce5e613221f69c8fcb7edbcf20a97b0c418d

mtreinish committed:
Apply suggestions from code review

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

mtreinish committed:
Fix issues identified during code review

This commit updates several issues found in the module during code
review.

mtreinish committed:
Update homeassistant/components/opnsense/__init__.py

Co-Authored-By: Fabian Affolter <mail@fabian-affolter.ch>

mtreinish committed:
Update CODEOWNERS for recent changes

mtreinish committed:
Fix lint

MartinHjelmare commented on tests/components/opnsense/test_device_tracker.py:
```diff
@@ -0,0 +1,41 @@
+async def test_get_scanner(hass):
+    """Test creating an opnsense scanner."""
+    with MockDependency("pyopnsense") as mocked_opnsense:
+        mocked_opnsense.diagnostic.InterfaceClient().get_arp = mock_coro_func()
```
We shouldn't set a mock coroutine as `get_arp` since it's not a coroutine function. Use a regular mock and have the return value of the mock return some fake data representing a seen device. Then we can assert the state of the entity at the end of the test.

mtreinish replied:
Fixed in: https://github.com/home-assistant/home-assistant/pull/26834/commits/f83386d9b87ddb3deec11414a6cb3bf955a457c5

MartinHjelmare commented on tests/components/opnsense/test_device_tracker.py:
```diff
@@ -0,0 +1,41 @@
+    """Test creating an opnsense scanner."""
+    with MockDependency("pyopnsense") as mocked_opnsense:
+        mocked_opnsense.diagnostic.InterfaceClient().get_arp = mock_coro_func()
+        mocked_opnsense.diagnostic.NetworkInsightClient().get_interfaces = mock_coro_func(
```
As above, this is not a coroutine.

mtreinish replied:
Fixed in: https://github.com/home-assistant/home-assistant/pull/26834/commits/f83386d9b87ddb3deec11414a6cb3bf955a457c5

mtreinish committed:
Apply suggestions from code review

Co-Authored-By: Martin Hjelmare <marhje52@kth.se>

mtreinish committed:
More fixes from review comments

This commit fixes several issues from review comments, including
abandoning all the use of async code. This also completely reworks the
tests to be a bit clearer.

mtreinish commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,70 @@
+    )
+    hass.data[OPNSENSE_DATA] = {"interfaces": interfaces_client}
+
+    hass.async_create_task(
```
Done in: https://github.com/home-assistant/home-assistant/pull/26834/commits/f83386d9b87ddb3deec11414a6cb3bf955a457c5

mtreinish commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,70 @@
+                )
+                return False
+    else:
+        tracker_interfaces = ["LAN"]
```
You're right I should change this logic. There's no guarantee an interface named `LAN` will exist on the router. It's just a default value when setting up opnsense. I will change the logic in device_tracker to use all devices unless it's specified. Fixed in: https://github.com/home-assistant/home-assistant/pull/26834/commits/f83386d9b87ddb3deec11414a6cb3bf955a457c5

MartinHjelmare commented on tests/components/opnsense/test_device_tracker.py:
```diff
@@ -1,41 +1,42 @@
-        CONF_VERIFY_SSL: False,
-    }
-}
+class TestOpnSenseDeviceTrackerSetup(unittest.TestCase):
```
The tests should still be standalone coroutine pytest test functions. Please just change the mocking as requested.

mtreinish replied:
Done in: https://github.com/home-assistant/home-assistant/pull/26834/commits/4f7c26932e865946feb6a812bc4f2d5ea6caf206

mtreinish committed:
Revert tests to previous format

mtreinish commented:
It looks like the linter job failed because of a network issue trying to talk to pypi: https://dev.azure.com/home-assistant/Home%20Assistant/_build/results?buildId=10835&view=logs&j=4e214fed-f310-53fa-0018-b38e7beb405b&t=065ca18d-6864-5af3-aac6-258b23c5c37b&l=87

MartinHjelmare commented on tests/components/opnsense/test_device_tracker.py:
```diff
@@ -0,0 +1,35 @@
+async def test_get_scanner(hass):
+    """Test creating an opnsense scanner."""
+    with MockDependency("pyopnsense") as mocked_opnsense:
+        get_arp = mock.MagicMock()
```
If we set the `return_value` for this mock to some mock device data with mac address etc, we can test that the state is updated correctly.

mtreinish replied:
Done in: https://github.com/home-assistant/home-assistant/pull/26834/commits/cd1bbe60d0140a59fdd547fc35db8b1747afb36a

mtreinish committed:
Add device detection to opnsense device_tracker test

This commit adds actual device detection to the unit test for the setup
test. A fake api response is added to mocks for both api clients so that
they will register devices as expected and asserts are added for that.

The pyopnsense import is moved from the module level to be runtime in
the class. This was done because it was the only way to make the
MockDependency() call work as expected.

mtreinish committed:
Rerun black

mtreinish committed:
Fix lint

MartinHjelmare commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,68 @@
+def setup(hass, config):
+    """Set up the opnsense component."""
+
+    from pyopnsense import diagnostics
```
Please move 3rd party imports to the top of the module.

mtreinish replied:
Done in: https://github.com/home-assistant/home-assistant/pull/26834/commits/d1d764b8ef25190603b2c85b43dc4e13d7f93451

But, see the commit message on: https://github.com/home-assistant/home-assistant/pull/26834/commits/cd1bbe60d0140a59fdd547fc35db8b1747afb36a (which made this change) it was the only way I managed to get the mocking working reliably. It only comes up now when I make the change you suggested in your other review comment.

MartinHjelmare commented on tests/components/opnsense/test_device_tracker.py:
```diff
@@ -0,0 +1,60 @@
+
+        result = await async_setup_component(
+            hass,
+            device_tracker.DOMAIN,
```
We should set up the opnsense component not the device tracker component. The opnsense component will load its device tracker platform.

mtreinish replied:
I had it this way originally (you can see earlier iterations) but found during local testing that this didn't happen until the test teardown. By making it manually load device_tracker here it happened during the test which is what we wanted. If you have a suggestion on how to make the test wait for the device tracker component to be setup during. Also when this is coupled to your above review the mock doesn't apply to all the http calls made in teardown and it leads to tracebacks from failing to connect

mtreinish replied:
With just this change (still with a runtime import):

```
------------------------------ Captured log call -------------------------------
INFO     homeassistant.loader:loader.py:212 Loaded test_package from custom_components.test_package
INFO     homeassistant.loader:loader.py:212 Loaded test from custom_components.test
INFO     homeassistant.loader:loader.py:212 Loaded opnsense from homeassistant.components.opnsense
INFO     homeassistant.setup:setup.py:145 Setting up opnsense
INFO     homeassistant.setup:setup.py:185 Setup of domain opnsense took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=opnsense>
INFO     homeassistant.loader:loader.py:212 Loaded device_tracker from homeassistant.components.device_tracker
---------------------------- Captured log teardown -----------------------------
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_stop[L]>
INFO     homeassistant.loader:loader.py:212 Loaded group from homeassistant.components.group
INFO     homeassistant.loader:loader.py:212 Loaded zone from homeassistant.components.zone
INFO     homeassistant.setup:setup.py:145 Setting up group
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=reload>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=remove>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set_visibility>
INFO     homeassistant.setup:setup.py:185 Setup of domain group took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=group>
INFO     homeassistant.setup:setup.py:145 Setting up zone
INFO     homeassistant.setup:setup.py:185 Setup of domain zone took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=zone>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=zone.home, old_state=None, new_state=<state zone.home=zoning; hidden=True, latitude=32.87336, longitude=-117.22743, radius=100, friendly_name=test home, icon=mdi:home @ 2019-10-02T19:12:43.197361+00:00>>
INFO     homeassistant.setup:setup.py:145 Setting up device_tracker
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=device_tracker, service=see>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event call_service[L]: domain=group, service=set, service_data=object_id=all_devices, visible=False, name=all devices, entities=['device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop']>
DEBUG    homeassistant.helpers.restore_state:restore_state.py:84 Not creating cache - no saved states found
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.ff_ff_ff_ff_ff_ff, old_state=None, new_state=<state device_tracker.ff_ff_ff_ff_ff_ff=not_home; source_type=None, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:12:43.208801+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.desktop, old_state=None, new_state=<state device_tracker.desktop=not_home; source_type=None, friendly_name=Desktop @ 2019-10-02T19:12:43.209131+00:00>>
INFO     homeassistant.setup:setup.py:185 Setup of domain device_tracker took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=device_tracker>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event platform_discovered[L]: service=load_platform.device_tracker, platform=opnsense, discovered=[]>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, old_state=None, new_state=<state group.all_devices=not_home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:12:43.210215+00:00>>
INFO     homeassistant.components.device_tracker:setup.py:55 Setting up device_tracker.legacy
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.ff_ff_ff_ff_ff_ff, old_state=<state device_tracker.ff_ff_ff_ff_ff_ff=not_home; source_type=None, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:12:43.208801+00:00>, new_state=<state device_tracker.ff_ff_ff_ff_ff_ff=home; source_type=router, latitude=32.87336, longitude=-117.22743, gps_accuracy=0, scanner=OPNSenseDeviceScanner, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:12:43.214202+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, old_state=<state group.all_devices=not_home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:12:43.210215+00:00>, new_state=<state group.all_devices=home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:12:43.215051+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.desktop, old_state=<state device_tracker.desktop=not_home; source_type=None, friendly_name=Desktop @ 2019-10-02T19:12:43.209131+00:00>, new_state=<state device_tracker.desktop=home; source_type=router, latitude=32.87336, longitude=-117.22743, gps_accuracy=0, scanner=OPNSenseDeviceScanner, manufacturer=OEM, friendly_name=Desktop @ 2019-10-02T19:12:43.215551+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_close[L]>
```

mtreinish replied:
With this change and a module level import:

```
------------------------------ Captured log call -------------------------------
INFO     homeassistant.loader:loader.py:212 Loaded test_package from custom_components.test_package
INFO     homeassistant.loader:loader.py:212 Loaded test from custom_components.test
INFO     homeassistant.loader:loader.py:212 Loaded opnsense from homeassistant.components.opnsense
INFO     homeassistant.setup:setup.py:145 Setting up opnsense
INFO     homeassistant.setup:setup.py:185 Setup of domain opnsense took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=opnsense>
---------------------------- Captured log teardown -----------------------------
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_stop[L]>
INFO     homeassistant.loader:loader.py:212 Loaded group from homeassistant.components.group
INFO     homeassistant.loader:loader.py:212 Loaded zone from homeassistant.components.zone
INFO     homeassistant.setup:setup.py:145 Setting up group
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=reload>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=remove>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=group, service=set_visibility>
INFO     homeassistant.setup:setup.py:185 Setup of domain group took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=group>
INFO     homeassistant.setup:setup.py:145 Setting up zone
INFO     homeassistant.setup:setup.py:185 Setup of domain zone took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=zone>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=zone.home, old_state=None, new_state=<state zone.home=zoning; hidden=True, latitude=32.87336, longitude=-117.22743, radius=100, friendly_name=test home, icon=mdi:home @ 2019-10-02T19:23:57.596806+00:00>>
INFO     homeassistant.setup:setup.py:145 Setting up device_tracker
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event service_registered[L]: domain=device_tracker, service=see>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event call_service[L]: domain=group, service=set, service_data=object_id=all_devices, visible=False, name=all devices, entities=['device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop']>
DEBUG    homeassistant.helpers.restore_state:restore_state.py:84 Not creating cache - no saved states found
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.ff_ff_ff_ff_ff_ff, old_state=None, new_state=<state device_tracker.ff_ff_ff_ff_ff_ff=not_home; source_type=None, friendly_name=ff ff ff ff ff ff @ 2019-10-02T19:23:57.606378+00:00>>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=device_tracker.desktop, old_state=None, new_state=<state device_tracker.desktop=not_home; source_type=None, friendly_name=Desktop @ 2019-10-02T19:23:57.606559+00:00>>
INFO     homeassistant.setup:setup.py:185 Setup of domain device_tracker took 0.0 seconds.
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event component_loaded[L]: component=device_tracker>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event platform_discovered[L]: service=load_platform.device_tracker, platform=opnsense, discovered=[]>
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event state_changed[L]: entity_id=group.all_devices, old_state=None, new_state=<state group.all_devices=not_home; entity_id=('device_tracker.ff_ff_ff_ff_ff_ff', 'device_tracker.desktop'), order=0, auto=True, friendly_name=all devices, hidden=True @ 2019-10-02T19:23:57.607193+00:00>>
INFO     homeassistant.components.device_tracker:setup.py:55 Setting up device_tracker.legacy
DEBUG    urllib3.connectionpool:connectionpool.py:959 Starting new HTTPS connection (1): fake_host_fun:443
ERROR    homeassistant.core:core.py:148 Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connection.py", line 157, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/util/connection.py", line 61, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/lib64/python3.7/socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 672, in urlopen
    chunked=chunked,
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 376, in _make_request
    self._validate_conn(conn)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 994, in _validate_conn
    conn.connect()
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connection.py", line 334, in connect
    conn = self._new_conn()
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connection.py", line 169, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6b40b3ed0>: Failed to establish a new connection: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/connectionpool.py", line 720, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/urllib3/util/retry.py", line 436, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='fake_host_fun', port=443): Max retries exceeded with url: /api/diagnostics/interface/getArp (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6b40b3ed0>: Failed to establish a new connection: [Errno -2] Name or service not known'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mtreinish/git/home-assistant/homeassistant/components/device_tracker/setup.py", line 164, in async_device_tracker_scan
    found_devices = await scanner.async_scan_devices()
  File "/usr/lib64/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/mtreinish/git/home-assistant/homeassistant/components/opnsense/device_tracker.py", line 39, in scan_devices
    self.update_info()
  File "/home/mtreinish/git/home-assistant/homeassistant/components/opnsense/device_tracker.py", line 55, in update_info
    devices = self.client.get_arp()
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/pyopnsense/diagnostics.py", line 52, in get_arp
    return self._get('diagnostics/interface/getArp')
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/pyopnsense/client.py", line 53, in _get
    timeout=DEFAULT_TIMEOUT)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/mtreinish/git/home-assistant/.tox/py37/lib/python3.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='fake_host_fun', port=443): Max retries exceeded with url: /api/diagnostics/interface/getArp (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6b40b3ed0>: Failed to establish a new connection: [Errno -2] Name or service not known'))
DEBUG    homeassistant.core:core.py:575 Bus:Handling <Event homeassistant_close[L]>
```

MartinHjelmare replied:
We need to `await hass.async_block_till_done()` after component setup to let the platform load.

MartinHjelmare replied:
I would put the dependency mock etc in a pytest fixture function. That way it's applied during the whole test.

mtreinish replied:
Done in: https://github.com/home-assistant/home-assistant/pull/26834/commits/600cbd59ce571a9eb6b51ece5ef65e0361109280

mtreinish committed:
Move import back to module level

mtreinish committed:
Return false on configuration errors in setup

This commit updates the connection logic to return false if we're unable
to connect to the configured OPNsense API endpoint for any reason.
Previously we would not catch if an endpoint was incorrectly configured
until we first tried to use it. In this case it would raise an unhandled
exception. To handle this more gracefully this adds an api call early in
the setup and catches any exception raised by that so we can return
False to indicate the setup failed.

mtreinish committed:
Update tests

MartinHjelmare commented on homeassistant/components/opnsense/__init__.py:
```diff
@@ -0,0 +1,76 @@
+    )
+    try:
+        interfaces_client.get_arp()
+    except Exception:  # pylint: disable=broad-except
```
Please catch a specific exception raised by the library on credentials or connection failure.

It's good if the library raises different exceptions for different type of problems.

mtreinish replied:
It depends on the failure mode, the library will raise a `pyopnsense.exceptions.APIException` if the api returns an error, because of invalid credentials for example. But if there is a connection issue or something else that happens at a lower level like requests or urllib3 that will get raised verbatim. Which is why I used a broad catch here. It seemed like it would be better to catch any exception and log it rather than raise an unhandled exception from a lower layer.

MartinHjelmare replied:
Preferably the library should catch RequestException and raise a library specific error.

MartinHjelmare commented:
If we need the library in the tests we need to add the library to the list in `script/gen_requirements_all.py`.

https://github.com/home-assistant/home-assistant/blob/bb45bdd8dd937a9fbbb21def1da93299498a9290/script/gen_requirements_all.py#L44

mtreinish committed:
Add pyopnsense to test requirements

frenck commented:
Hi @mtreinish!

Looks like this PR is pretty close to the finish 🥇 
Are you planning to continue to work on this PR?

At this point, I would recommend rebasing the PR onto the latest dev branch, regenerate the requirements, and go from there.

Looking forward to it! 👍

mtreinish committed:
Rerun gen_requirements script

springstan commented:
Hello @mtreinish the failed tests seem to be unrelated to this PR, that is why I would recommend you to rebase onto the latest dev branch again. This way the tests have to rerun and hopefully pass 😅 so this can be finally merged ✌

mtreinish committed:
Fix failing isort lint job step

Since opening the PR originally yet another lint/style checker was added
which failed the PR in CI. This commit makes the adjustments to have
this pass the additional tool's checks.

pvizeli committed:
Fix comment

pvizeli committed:
Update manifest.json
